src/main/java/org/dave/compactmachines3/world/WorldSavedDataMachines.java

```diff
@@ -1,131 +1,198 @@
+        for (Map.Entry<Integer, BlockPos> entry : machineGrid.entrySet()) {
+            int roomPosX = entry.getValue().getX();
+            int roomPosZ = entry.getValue().getZ();
+            EnumMachineSize sizeEnum = machineSizes.get(entry.getKey());
+            if (sizeEnum == null) {
+                CompactMachines3.logger.error("Machine size was null with key {}", entry.getKey());
+                continue;
+            }
+            int size = sizeEnum.getDimension();
+            boolean insideRoom = roomPosX <= x && x <= roomPosX + size && roomPosZ <= z && z <= roomPosZ + size && 40 <= y && y <= 40 + size;
+
+            if (insideRoom) {
+                return entry.getKey();
+            }
         }
 
-        spawnPoints.put(coord, destination);
-        Logz.debug("Setting spawn point: coords=%d, x=%.2f, y=%.2f, z=%.2f", coord, destination[0], destination[1], destination[2]);
+        return -1;
+    }
+
+    public void addMachineSize(int id, EnumMachineSize size) {
+        machineSizes.put(id, size);
+        CompactMachines3.logger.debug("Adding machine size: id={}, size={}", id, size.getName());
         this.markDirty();
     }
 
-    public void addSpawnPoint(int coord, double x, double y, double z) {
-        addSpawnPoint(coord, new double[]{x, y, z});
+    public void addMachinePosition(int id, BlockPos pos, int dimension) {
+        machinePositions.put(id, new DimensionBlockPos(pos, dimension));
+        CompactMachines3.logger.debug("Adding machine position: id={}, pos={}, dimension={}", id, pos, dimension);
+        this.markDirty();
+    }
+
+    public void addSpawnPoint(int id, @Nonnull Vec3d destination) {
+        spawnPoints.put(id, destination);
+        CompactMachines3.logger.debug(String.format("Setting spawn point: id=%s, x=%.2f, y=%.2f, z=%.2f", id, destination.x, destination.y, destination.z));
+        this.markDirty();
     }
 
-    public void removeMachinePosition(int coord) {
-        machinePositions.remove(coord);
-        Logz.debug("Removing machine position by coord: coords=%d", coord);
+    public void removeMachinePosition(int id) {
+        machinePositions.remove(id);
+        CompactMachines3.logger.debug("Removing machine position by id: id={}", id);
         this.markDirty();
     }
 
     // TODO: We might want to move this to a different class?
     public static int reserveMachineId() {
-        int val = WorldSavedDataMachines.INSTANCE.nextCoord;
-        WorldSavedDataMachines.INSTANCE.nextCoord++;
-        WorldSavedDataMachines.INSTANCE.markDirty();
+        WorldSavedDataMachines wsd = WorldSavedDataMachines.getInstance();
+        int val = wsd.nextId++; // Returns the value, then adds 1
+        wsd.markDirty();
         return val;
     }
 
     public WorldSavedDataMachines(String name) {
         super(name);
     }
 
-    @SubscribeEvent
+    @SubscribeEvent(priority = EventPriority.HIGHEST)
```

Comment:
Why the priority increase?

Reply:
Not sure, probably just to load the WSDM as soon as possible so as to avoid other race conditions. Probably an unnecessary change.